k8s/upgrade/src/preflight_validations.rs

```diff
@@ -1,26 +1,147 @@
+    timeout: humantime::Duration,
+) -> Result<(), error::Error> {
     console_logger::info(user_prompt::UPGRADE_WARNING);
-    rebuild_in_progress_validation().await?;
+    // Initialise the REST client.
+    let config = kube_proxy::ConfigBuilder::default_api_rest()
+        .with_kube_config(kube_config_path.clone())
+        .with_timeout(*timeout)
+        .build()
+        .await?;
+
+    let rest_client = upgrade_lib::RestClient::new_with_config(config);
+
+    rebuild_in_progress_validation(&rest_client).await?;
+
+    already_cordoned_nodes_validation(&rest_client).await?;
+
+    single_volume_replica_validation(&rest_client).await?;
+    Ok(())
+}
+
+/// Check single replica volumes.
+pub async fn already_cordoned_nodes_validation(
+    client: &upgrade_lib::RestClient,
+) -> Result<(), error::Error> {
+    let mut cordoned_nodes_list = Vec::new();
+    let nodes = client.nodes_api().get_nodes().await?;
+    let nodelist = nodes.into_body();
+    for node in nodelist {
+        let node_spec = match node.spec {
+            Some(node_spec) => node_spec,
+            None => return Err(error::Error::NodeSpecNotPresent { node: node.id }),
+        };
+        match node_spec.cordondrainstate {
+            Some(CordonDrainState::cordonedstate(_)) => {
+                cordoned_nodes_list.push(node.id);
+            }
+            Some(CordonDrainState::drainingstate(_)) => (),
+            Some(CordonDrainState::drainedstate(_)) => (),
+            None => (),
+        };
+    }
+    if !cordoned_nodes_list.is_empty() {
+        console_logger::warn(
+            user_prompt::CORDONED_NODE_WARNING,
+            &cordoned_nodes_list.join("\n"),
+        );
+    }
+    Ok(())
+}
+
+/// Check single replica volumes.
+pub async fn single_volume_replica_validation(
+    client: &upgrade_lib::RestClient,
+) -> Result<(), error::Error> {
+    // let mut single_replica_volumes = Vec::new();
+    // The number of volumes to get per request.
+    let max_entries = 200;
+    let mut starting_token = Some(0_isize);
+    let mut volumes = Vec::with_capacity(max_entries as usize);
+
+    // The last paginated request will set the `starting_token` to `None`.
+    while starting_token.is_some() {
+        let vols = client
+            .volumes_api()
+            .get_volumes(max_entries, None, starting_token)
+            .await
+            .map_err(|error| {
+                error!(?error, "Failed to list volumes");
+                error
+            })?;
+
+        let v = vols.into_body();
+        let single_rep_vol_ids: Vec<String> = v
+            .entries
+            .into_iter()
+            .filter(|volume| volume.spec.num_replicas == SINGLE_REPLICA_VOLUME)
+            .map(|volume| volume.spec.uuid.to_string())
+            .collect();
+        volumes.extend(single_rep_vol_ids);
+        starting_token = v.next_token;
+    }
+
+    let single_volume_replicas_list: Vec<String> = volumes.iter().map(|s| s.to_string()).collect();
```

Comment:
Why do we again need to convert to string? The `volumes` vector already has the String right?

Reply:
ah, thanks removed it